### PR TITLE
Add a n_startup_trials option to SkoptSampler.

### PR DESCRIPTION
**Pull request by c-bata**
Provide a `n_startup_trials` option like CmaEsSampler.

### CODE REVIEW

**Comment by c-bata** [comment]
I re-run the workflow of CircleCI, then passed all checks.

CircleCI seems to be triggered on my private account. I don't know why but probably this problem was resolved by unfollowing `c-bata/optuna` project on my CircleCI settings.

ScreenShot 2020-02-26 17 37 45

**Comment by toshihikoyanase** [comment]
@c-bata 
I found that your PR was tested in both optuna account and c-bata account.
- https://circleci.com/gh/optuna/optuna/3509
- https://circleci.com/gh/c-bata/optuna/3509

I guess you have webhook settings of CircleCI in your repository [here](https://github.com/c-bata/optuna/settings/hooks).  If so, is it possible to clear the settings?
 I don't set any webhook settings for https://github.com/toshihikoyanase/optuna, but CI jobs are successfully triggered by https://github.com/optuna/optuna.

**Comment by c-bata** [comment]
Hi @toshihikoyanase. Thank you for your investigation.
I've just checked it, but there are no webhook settings there.

ScreenShot 2020-02-26 18 11 05

**Comment by toshihikoyanase** [comment]
@c-bata 
Thank you for your quick response.
The webhook settings are not related to this issue.
I'll investigate further and I'll let you know if I have any clues.

**Comment by grassking100** [comment]
Hello,
      If I understand correctly, the "n_startup_trials" is used to determine whether to use independent sampler or not. If "n_startup_trials" is smaller than the number of complete trials then it would use sample_relative, otherwise, it would use sample_independent.
     But there is an option named "n_initial_points" in the skopt sampler [[link](https://scikit-optimize.github.io/stable/modules/generated/skopt.optimizer.Optimizer.html?highlight=n_initial_points)], which does the same thing. I wonder this would cause some ambiguity. Maybe we should print the message to warn the user that "n_initial_points" in the skopt sampler won't be used. Or we should remove the "n_startup_trials" option to eliminate the ambiguity.

**Comment by c-bata** [comment]
Hi @grassking100. Thank you for your comment. The purpose of `n_startup_trials` is actually different with `n_initial_points`.

Optuna adopts a define-by-run interface. This means we cannot obtain the search_space until executing objective functions. `n_starup_trials` option is used to determine the search space. On the other hand, `n_initial_points` of scikit-optimize is used for building a surrogate model.

**Comment by grassking100** [comment]
Hi @c-bata. Thank you for your reply. 

    def sample_relative(self, study, trial, search_space):
        # type: (Study, FrozenTrial, Dict[str, BaseDistribution]) -> Dict[str, Any]

        if len(search_space) == 0:
            return {}

        complete_trials = [t for t in study.trials if t.state == structs.TrialState.COMPLETE]
        if len(complete_trials) < self._n_startup_trials:
            return {}

        optimizer = _Optimizer(search_space, self._skopt_kwargs)
        optimizer.tell(study, complete_trials)
        return optimizer.ask()


If I understand the above code correctly, the **n_starup_trials** would decide whether to ask the parameters from the scikit-optimizer or not. If the **n_starup_trials** is 5 and the number of completed trials is 1, then the scikit-optimizer. wouldn't be asked, and the parameters would be decided by sample_independent.  If the **n_starup_trials** is 5 and the number of completed trials is 7, then the scikit-optimizer would be asked, and the parameters would be decided by scikit-optimizer.

**Comment by c-bata** [comment]
> If I understand the above code correctly, the n_starup_trials would decide whether to ask the parameters from the scikit-optimizer or not. If the n_starup_trials is 5 and the number of completed trials is 1, then the scikit-optimizer. wouldn't be asked, and the parameters would be decided by sample_independent. If the n_starup_trials is 5 and the number of completed trials is 7, then the scikit-optimizer would be asked, and the parameters would be decided by scikit-optimizer.

Yes, your recognition is right.

I'm sorry. I don't know what you want to ask.

**Comment by grassking100** [comment]
Hello @c-bata :
So if the n_starup_trials is 5 and the number of completed trials is 7, then we would expect **sample_relative** use the scikit-optimizer to sample parameter by its base_estimator but not randomly. But the default **n_initial_points** in the scikit-optimizer is set to 10, so the scikit-optimizer would choose parameter randomly and independently [[link](https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/optimizer/optimizer.py#L423)], and this is not we expect what **sample_relative** would do. Maybe we should set **n_initial_points** to zero like the following code, so the method **sample_relative** would work as we expect.

```diff
        def __init__(self, independent_sampler=None, warn_independent_sampling=True,
                 skopt_kwargs=None, n_startup_trials=1):
            #type: (Optional[BaseSampler], bool, Optional[Dict[str, Any]], int) -> None

             _check_skopt_availability()
 
            self._skopt_kwargs = skopt_kwargs or {}
            if 'dimensions' in self._skopt_kwargs:
                del self._skopt_kwargs['dimensions']

+++         self._skopt_kwargs['n_initial_points'] = 0

            self._independent_sampler = independent_sampler or samplers.RandomSampler()
            self._warn_independent_sampling = warn_independent_sampling
            self._n_startup_trials = n_startup_trials
```

**Comment by grassking100** [comment]
> > If I understand the above code correctly, the n_starup_trials would decide whether to ask the parameters from the scikit-optimizer or not. If the n_starup_trials is 5 and the number of completed trials is 1, then the scikit-optimizer. wouldn't be asked, and the parameters would be decided by sample_independent. If the n_starup_trials is 5 and the number of completed trials is 7, then the scikit-optimizer would be asked, and the parameters would be decided by scikit-optimizer.
> 
> Yes, your recognition is right.
> 
> I'm sorry. I don't know what you want to ask.
Hello @c-bata :
    How do you think of my suggestion in the above comment? Should we enforce the 'n_initial_points' to be zero, or should we warn the user that if it is not zero, it would cause some unexpected behavior?

**Comment by grassking100** [comment]
Hello @c-bata :
I wrote some code to do some demonstrations. I set simple search space param_loss between -100 to 0. The objective function would return the "-param_loss". The independent sampler is DeterministicRelativeSampler which uses the lowest value to be the hyperparameter. 

**I optimize the hyperparameter with 20 trials, the first 5 trials would be 5. I would repeat it 10 times and record its loss summation at each epoch. I expect the loss values would stay at 100 at the first 5 trials, and the values would decrease to zero.** 

In example 1, skopt_kwargs would use their default value. In the plot of example 1, we could see values would stay at 100 in the first 5 trials, and values would decrease to 500 in epoch 6~10, then the values would decrease to 0.


In example 2, the n_initial_points would be 0. In the plot of example 2, we could see values would stay at 100 in the first 5 trials, and the values would decrease to 0.

Maybe we should set **n_initial_points** to be zero, so the method sample_relative would work as we expect.

example_1

example_2



```python
import optuna
import numpy as np
from optuna.testing.sampler import DeterministicRelativeSampler
from matplotlib import pyplot as plt

def objective(trial):
    loss = trial.suggest_int("param_loss",-100,0)
    return -loss

def default_sampler_example():
    independent_sampler = DeterministicRelativeSampler({}, {})
    skopt_kwargs = {'random_state': 0}
    sampler = optuna.integration.SkoptSampler(n_startup_trials=5,
                                              independent_sampler=independent_sampler,
                                              skopt_kwargs=skopt_kwargs)
    return sampler
    
def suggested_sampler_example():
    independent_sampler = DeterministicRelativeSampler({}, {})
    skopt_kwargs = {'random_state': 0,'n_initial_points': 0}
    sampler = optuna.integration.SkoptSampler(n_startup_trials=5,
                                              independent_sampler=independent_sampler,
                                              skopt_kwargs=skopt_kwargs)
    return sampler

def optimize_with_sampler(sampler):
    loss_sum = None
    for _ in range(10):
        study = optuna.create_study(sampler=sampler)
        study.optimize(objective, n_trials=20)
        loss = np.array(study.trials_dataframe()['value'])
        if loss_sum is None:
            loss_sum = loss
        else:
            loss_sum += loss
    return loss_sum

#Example 1
loss = optimize_with_sampler(default_sampler_example())
plt.title("Unexpect training loss curve in example 1")
plt.xlabel('trial')
plt.ylabel('loss summation')
plt.show()

#Example 2
loss2 = optimize_with_sampler(suggested_sampler_example())
plt.title("Expect training loss curve in example 2")
plt.xlabel('trial')
plt.ylabel('loss summation')
plt.show()
```
